base/dates/parse.jl

```diff
@@ -1,97 +1,132 @@
 ### Parsing utilities
 
-@generated function tryparse_internal{T<:TimeType, S, F}(::Type{T}, str::AbstractString, df::DateFormat{S, F}, raise::Bool=false)
-    token_types = Type[dp <: DatePart ? SLOT_RULE[first(dp.parameters)] : Void for dp in F.parameters]
-    N = length(F.parameters)
-
-    types = slot_order(T)
-    num_types = length(types)
-    order = Vector{Int}(num_types)
-    for i = 1:num_types
-        order[i] = findfirst(token_types, types[i])
+function directives{S,F}(::Type{DateFormat{S,F}})
+    tokens = F.parameters
+    di = 1
+    directive_index = zeros(Int, length(tokens))
+    directive_letters = sizehint!(Char[], length(tokens))
+    for (i, token) in enumerate(tokens)
+        if token <: DatePart
+            directive_index[i] = di
+
+            letter = first(token.parameters)
+            push!(directive_letters, letter)
+
+            di += 1
+        end
     end
+    return tokens, directive_index, directive_letters
+end
+
+genvar(t::DataType) = Symbol(lowercase(string(t.name.name)))
 
-    field_defaults = slot_defaults(T)
-    field_order = tuple(order...)
-    tuple_type = slot_types(T)
 
-    # `slot_order`, `slot_defaults`, and `slot_types` return tuples of the same length
-    assert(num_types == length(field_order) == length(field_defaults))
+@generated function tryparse_core(str::AbstractString, df::DateFormat, raise::Bool=false)
```

Comment:
It will be great if this function can take a start index and return an end position... I've had to hack around this to get this, because it's very useful when you're parsing dates and some more stuff after / a list of dates